plugin/storage/badger/spanstore/reader.go

```diff
@@ -155,69 +168,55 @@ func (r *TraceReader) GetTrace(ctx context.Context, traceID model.TraceID) (*mod
 }
 
 // scanTimeRange returns all the Traces found between startTs and endTs
-func (r *TraceReader) scanTimeRange(startTime time.Time, endTime time.Time) ([]*model.Trace, error) {
+func (r *TraceReader) scanTimeRange(plan *executionPlan) ([]model.TraceID, error) {
 	// We need to do a full table scan
-	traces := make([]*model.Trace, 0)
-	startTs := model.TimeAsEpochMicroseconds(startTime)
-	endTs := model.TimeAsEpochMicroseconds(endTime)
-
+	traceKeys := make([][]byte, 0)
 	err := r.store.View(func(txn *badger.Txn) error {
 		opts := badger.DefaultIteratorOptions
-		opts.PrefetchValues = true
+		opts.PrefetchValues = false
 		it := txn.NewIterator(opts)
 		defer it.Close()
 
-		val := []byte{}
 		startIndex := []byte{spanKeyPrefix}
-		spans := make([]*model.Span, 0)
 		prevTraceID := []byte{}
 		for it.Seek(startIndex); it.ValidForPrefix(startIndex); it.Next() {
 			item := it.Item()
 
 			key := []byte{}
 			key = item.KeyCopy(key)
 
-			timestamp := binary.BigEndian.Uint64(key[sizeOfTraceID+1 : sizeOfTraceID+1+8])
+			timestamp := key[sizeOfTraceID+1 : sizeOfTraceID+1+8]
 			traceID := key[1 : sizeOfTraceID+1]
 
-			if timestamp >= startTs && timestamp <= endTs {
-				val, err := item.ValueCopy(val)
-				if err != nil {
-					return err
-				}
-
-				sp, err := decodeValue(val, item.UserMeta()&encodingTypeBits)
-				if err != nil {
-					return err
-				}
-
-				if bytes.Equal(prevTraceID, traceID) {
-					// Still processing the same one
-					spans = append(spans, sp)
-				} else {
-					// Process last complete span
-					trace := &model.Trace{
-						Spans: spans,
+			if bytes.Compare(timestamp, plan.startTimeMin) >= 0 && bytes.Compare(timestamp, plan.startTimeMax) <= 0 {
+				if !bytes.Equal(traceID, prevTraceID) {
+					if plan.hashOuter != nil {
+						trID := bytesToTraceID(traceID)
+						if _, exists := plan.hashOuter[trID]; exists {
+							traceKeys = append(traceKeys, key)
+						}
+					} else {
+						traceKeys = append(traceKeys, key)
 					}
-					traces = append(traces, trace)
-
-					spans = make([]*model.Span, 0, cap(spans)) // Use previous cap
-					spans = append(spans, sp)
+					prevTraceID = traceID
 				}
-				prevTraceID = traceID
 			}
 		}
-		if len(spans) > 0 {
-			trace := &model.Trace{
-				Spans: spans,
-			}
-			traces = append(traces, trace)
-		}
 
 		return nil
 	})
 
-	return traces, err
+	sort.Slice(traceKeys, func(k, h int) bool {
+		return bytes.Compare(traceKeys[k][sizeOfTraceID+1:sizeOfTraceID+1+8], traceKeys[h][sizeOfTraceID+1:sizeOfTraceID+1+8]) > 0
```

Comment:
I think this deserves a comment, like:

    // compares traceKeys[k][timestamp] with traceKeys[h][timestamp]